src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -541,29 +557,59 @@ public WorkflowActivationCompletion Activate(WorkflowActivation act)
-                                    return 2;
-                                case WorkflowActivationJob.VariantOneofCase.InitializeWorkflow:
-                                    return 3;
-                                default:
-                                    return 4;
-                            }
-                        }).ToList();
-                        // We can trust jobs are deterministically ordered by core
-                        foreach (var job in sortedJobs)
+                                switch (j.VariantCase)
+                                {
+                                    case WorkflowActivationJob.VariantOneofCase.NotifyHasPatch:
+                                    case WorkflowActivationJob.VariantOneofCase.UpdateRandomSeed:
+                                        return 1;
+                                    case WorkflowActivationJob.VariantOneofCase.SignalWorkflow:
+                                    case WorkflowActivationJob.VariantOneofCase.DoUpdate:
+                                        return 2;
+                                    case WorkflowActivationJob.VariantOneofCase.InitializeWorkflow:
+                                        return 3;
+                                    default:
+                                        return 4;
+                                }
+                            }).ToList();
+                        }
+
+                        // Apply each job
+                        foreach (var job in jobs)
                         {
                             Apply(job);
-                            // Run scheduler once. Do not check conditions when patching or querying.
-                            var checkConditions = job.NotifyHasPatch == null && job.QueryWorkflow == null;
+                            // We only run the scheduler after each job in legacy event loop logic
+                            if (!applyModernEventLoopLogic)
+                            {
+                                // Run scheduler once. Do not check conditions when patching or
+                                // querying with legacy event loop logic.
+                                var checkConditions = job.NotifyHasPatch == null && job.QueryWorkflow == null;
+                                RunOnce(checkConditions);
+                            }
+                        }
+
+                        // For modern event loop logic, we initialize here if not initialized
+                        // already
+                        if (applyModernEventLoopLogic && !workflowInitialized)
+                        {
+                            InitializeWorkflow();
+                        }
+
+                        // For modern event loop logic, we run the event loop only after applying
+                        // everything, and we check conditions if there are any non-query jobs
+                        if (applyModernEventLoopLogic)
```

Comment:
nit:
```
// For modern event loop logic
if (applyModernEventLoopLogic) {
    // We initialize here if not initialized already
    if (!workflowInitialized) { ... }
    // We run the event loop only after applying everything, and we check conditions if there are any non-query jobs.
    var checkConditions = jobs.Any(j => j.VariantCase != WorkflowActivationJob.VariantOneofCase.QueryWorkflow);
    RunOnce(checkConditions); 
}
```
avoid the double check for `applyModernEventLoopLogic`, not a strong opinion, feel free to ignore

Reply:
I considered combining, but the legibility of separating initialize from run-once is worth the top-level separation